packages-ts/gauntlet-terra-contracts/src/lib/contracts.ts

```diff
@@ -28,109 +30,191 @@ export type TerraABI = {
-  ]
-
-  const toDirName = (contractId: CONTRACT_LIST) => contractDirName[contractId]
-
-  const abi = possibleContractPaths
-    .filter((path) => existsSync(`${path}/${toDirName(contractId)}/schema`))
-    .map((contractPath) => {
-      const toPath = (type) => {
-        if (contractId == CONTRACT_LIST.CW20_BASE && type == 'execute_msg') {
-          return path.join(contractPath, `./${toDirName(contractId)}/schema/cw20_${type}`)
-        } else {
-          return path.join(contractPath, `./${toDirName(contractId)}/schema/${type}`)
-        }
-      }
-      return {
-        execute: io.readJSON(toPath('execute_msg')),
-        query: io.readJSON(toPath('query_msg')),
-        instantiate: io.readJSON(toPath('instantiate_msg')),
-      }
-    })
-  if (abi.length === 0) {
-    logger.error(`ABI not found for contract ${contractId}`)
+
+class Contracts {
+  // Retrieves a specific Contract object from the contract index, while loading its abi
+  // and bytecode from disk or network if they haven't been already.
+  async getContractWithSchemaAndCode(id: CONTRACT_LIST, version: string): Promise<Contract> {
+    const contract = this[id]
+    if (!contract) {
+      throw new Error(`Contract ${id} not found!`)
+    }
+    await Promise.all([
+      contract.abi ? Promise.resolve() : contract.loadContractABI(),
+      contract.bytecode ? Promise.resolve() : contract.loadContractCode(version),
+    ])
+    return contract
   }
 
-  return abi[0]
+  addChainlink = (id: CONTRACT_LIST, dirName: string) => {
+    this[id] = new ChainlinkContract(id, dirName)
+    return this
+  }
+
+  addCosmwasm = (id: CONTRACT_LIST, dirName: string) => {
+    this[id] = new CosmWasmContract(id, dirName)
+    return this
+  }
+}
+
+export const contracts = new Contracts()
+  .addChainlink(CONTRACT_LIST.FLAGS, 'flags')
+  .addChainlink(CONTRACT_LIST.DEVIATION_FLAGGING_VALIDATOR, 'deviation-flagging-validator')
+  .addChainlink(CONTRACT_LIST.OCR_2, 'ocr2')
+  .addChainlink(CONTRACT_LIST.PROXY_OCR_2, 'proxy-ocr2')
+  .addChainlink(CONTRACT_LIST.ACCESS_CONTROLLER, 'access-controller')
+  .addCosmwasm(CONTRACT_LIST.CW20_BASE, 'cw20_base')
+  .addCosmwasm(CONTRACT_LIST.CW4_GROUP, 'cw4_group')
+  .addCosmwasm(CONTRACT_LIST.MULTISIG, 'cw3_flex_multisig')
+
+type Instance = {
+  name: string
+  contractId: CONTRACT_LIST
 }
 
-export const getContract = async (id: CONTRACT_LIST, version): Promise<Contract> => {
-  version = version ? version : defaultContractVersions[id]
-  // Preload contracts
-  return {
-    id,
-    abi: getContractABI(id),
-    bytecode: await getContractCode(id, version),
+class AddressBook {
+  operator: AccAddress
+  instances: Map<AccAddress, Instance> // address => instance name
+
+  constructor() {
+    this.instances = new Map<AccAddress, Instance>()
+  }
+
+  addOperator(address: AccAddress) {
+    this.operator = address
+  }
+  addInstance(contractId: CONTRACT_LIST, envVar: string, name?: string) {
+    const address = process.env[envVar]
```

Comment:
I was thinking we should get away from the practice of requiring each type of command to do its own loading of environment vars.  Why have the loading and validation of each contract addresses scattered throughout the codebase, when it can all be done in one centralized place, which all of the commands have access to?

I see your point that we should keep it as general as possible, to make it easier later when we switch over to RDD or other sources instead of the environment.

I can move the loading & validation of addresses out of the `AddressBook` class itself, into the middleware component which adds the entries to the AddressBook.  Does that work?